server/src/main/java/io/crate/metadata/Reference.java

```diff
@@ -83,6 +89,32 @@ static int indexOf(Iterable<? extends Reference> refs, ColumnIdent column) {
      */
     Map<String, Object> toMapping(int position);
 
+    @Nullable
+    default IndexableFieldType luceneFieldType() {
+        var name = column().name();
+        var valueType = valueType();
+        if (valueType.id() == IntegerType.ID) {
+            if (hasDocValues() && indexType() != IndexType.NONE) {
+                return new IntField(name, 0, Field.Store.NO).fieldType();
+            } else {
+                if (indexType() != IndexType.NONE) {
+                    return new IntPoint(name, 0).fieldType();
+                }
+                if (hasDocValues()) {
+                    return new SortedNumericDocValuesField(name, 0).fieldType();
+                }
+            }
+        }
+        return null;
+    }
+
+    default boolean isFieldNameIndexed() {
+        if (valueType().id() == IntegerType.ID) {
+            return !hasDocValues();
+        }
+        return false;
+    }
+
```

Comment:
This flag does not work as intended since the `FieldType` used when creating a column and when indexing a value are different and the one needed for `IsNullPredicate` is the latter. Also, to my understanding obtaining the `FieldType` is a deterministic process, hence `Reference.luceneFieldType()`, `Reference.isFieldNameIndexed()` are introduced.

Reply:
If this is accepted, there should be a follow up task to remove columns' `FieldTypes` persisted in cluster state.

Reply:
As you said, the field types are dictated by the value indexers and DocumentParser. Both will detect the used field type based on the cluster state. We could streamline this to one place, but this should be type specific (avoid any switch which must be maintained despite of the type system and such is error prone). 
I do not completely understand why we cannot use the lucene field-type for the decisions, could you please elaborate on this?

> If this is accepted, there should be a follow up task to remove columns' FieldTypes persisted in cluster state

I don't quite understand this, what exactly do you mean?

Reply:
I [fixed `canUseFieldsExist` flag](https://github.com/crate/crate/pull/14604#discussion_r1320891131)
but the test still fails because the `FieldType` persisted in the cluster state was wrong. The cause seems to be that when we create a column, we use default `FieldTypes` such as `NumberFieldMapper.FIELD_TYPE` that do not reflect on `index off` or `columnstore=false`. Then when we index values, the value indexers generate lucene fields based on the index/doc-value where the lucene fields contain proper `FieldTypes` but not persisted.
 
 My last commit would regenerate lucene field used for indexing and extract the `FieldType` instead of what was persisted as part of the column creation.

Reply:
> 
> > If this is accepted, there should be a follow up task to remove columns' FieldTypes persisted in cluster state
> 
> I don't quite understand this, what exactly do you mean?

Currently, I do not see the real use for the default `FieldTypes` statically declared in each `FieldMappers` that gets persisted. I was just thinking if we can avoid persisting at all.

Reply:
We definitely shouldn't add logic specific to a concrete type to `Reference` and if we were to add `FieldType` logic to `Reference` we also need a plan about what we're going to do with the `FieldMapper` infrastructure.

>  I do not see the real use for the default FieldTypes statically declared in each FieldMappers that gets persisted. I was just thinking if we can avoid persisting at all.


The default FieldTypes are used as template/default value when a user creates a table. We need to persist the information somehow, otherwise how else would you know which columns were created with doc-values vs. without?

Reply:
>The default FieldTypes are used as template/default value when a user creates a table. We need to persist the information somehow, otherwise how else would you know which columns were created with doc-values vs. without?

At least for doc-values and indexing they are persisted in `MappingMetadata`. If I trace
`create table t (a int index off storage with (columnstore=false), b int)` the following will be persisted to the cluster state:
```
{"default":{"dynamic":"strict","_meta":{},"properties":{"a":{"type":"integer","index":false,"doc_values":false,"position":1},"b":{"type":"integer","position":2}}}}
```

and `FieldTypes` available in `MappingService` for column `a` and `b` are identical.

Reply:
reverted, this is not an issue anymore.